tests/test_data_processing.py

```diff
@@ -26,14 +38,512 @@
+
+    # Time series
+    path_file_timeseries = os.path.join(
+        os.path.abspath(os.path.join(this_path, os.pardir)),
+        "_files",
+        "test_timeseries.csv",
+    )
+    df = load_timeseries(path_file_timeseries)
+
+    path_file_stacked = os.path.join(
+        os.path.abspath(os.path.join(this_path, os.pardir)),
+        "_files",
+        "test_stacked_saved.csv",
+    )
+
+    save_df(df, path_file_stacked)
+    assert os.path.exists(path_file_stacked) == 1
+
+    df_saved = load_timeseries(path_file_stacked)
+
+    pd.testing.assert_series_equal(df["series"], df_saved["series"])
+    pd.testing.assert_frame_equal(df, df_saved)
+
+    os.remove(path_file_stacked)
+
+
+def test_filter_df():
+    """
+    This test checks whether
+    1. scalars and
+    2. time series
+    are filtered by a key and value
+
+    """
+    # 1. Test
+    path_file_scalars = os.path.join(
+        os.path.abspath(os.path.join(this_path, os.pardir)),
+        "_files",
+        "test_scalars.csv",
+    )
+
+    # Read scalars
+    df = load_scalars(path_file_scalars)
+
+    df_BE = filter_df(df, "region", ["BE"])
+
+    assert df_BE["region"].values[0] == "BE"
+    assert len(df_BE["region"]) == 7
+
+    df_BE_BB = filter_df(df, "region", ["BE_BB"])
+
+    assert df_BE_BB.empty
+
+    df_conversion = filter_df(df, "type", ["conversion"])
+    assert len(df_conversion["type"]) == 10
+
+    with pytest.raises(KeyError):
+        filter_df(df, "something", ["conversion"])
+
+    # 2. Test
+    path_file_timeseries = os.path.join(
+        os.path.abspath(os.path.join(this_path, os.pardir)),
+        "_files",
+        "test_stacked.csv",
+    )
+
+    # Read stacked time series
+    df = load_timeseries(path_file_timeseries)
+
+    df_BE_BB = filter_df(df, "region", ["BE_BB"])
+
+    assert (
+        df_BE_BB["region"].values[0] == "BE_BB"
+        and df_BE_BB["region"].values[1] == "BE_BB"
+    )
+    assert len(df_BE_BB["region"]) == 2
+
+    df_var_name = filter_df(
+        df,
+        "var_name",
+        [
+            "flow from BB-biomass-st to BB-electricity",
+            "flow from BB-electricity-liion-battery to BB-electricity",
+        ],
+    )
+
+    assert (
+        df_var_name["var_name"].values[0] == "flow from BB-biomass-st to BB-electricity"
+        and df_var_name["var_name"].values[1]
+        == "flow from BB-electricity-liion-battery to BB-electricity"
+    )
+
+
+def test_df_agg():
+    """
+    This test checks whether
+    1. scalars and
+    2. time series
+    are aggregated by a key
+
+    """
+    # 1. Test
+    path_file_scalars = os.path.join(
+        os.path.abspath(os.path.join(this_path, os.pardir)),
+        "_files",
+        "test_scalars.csv",
+    )
+
+    # Read scalars
+    df = load_scalars(path_file_scalars)
+
+    # Aggregate by region
+    df_aggregated_by_region = df_agg(df, "region")
+
+    # Add expected results of aggregation
+    df_aggregated_expected = pd.DataFrame(
```

Comment:
Please save this as a .csv under _files.

Reply:
Done with commit https://github.com/rl-institut/oemof-B3/pull/57/commits/10429062c7180571172a6d83d4628f3da593b5ba.